### PR TITLE
Replace console.log calls with debug

### PR DESCRIPTION
**Pull request by mikehazell**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Replaces all console logs with logging calls via [debug](https://github.com/visionmedia/debug)

* **What is the current behavior?** (You can also link to an open issue here)

The console is full of unformatted logs which are noisy and hard to filter out. #901 

* **What is the new behavior (if this is a feature change)?**

A beautifully silent console unless:

1. An error condition occurs
2. The user enables log output

```js
localStorage.setItem('debug', 'cornerstoneTools:*');
``` 

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: src/eventDispatchers/touchEventHandlers/addNewMeasurement.js [hunk 1]
```diff
@@ -7,13 +7,18 @@ import {
   addToolState,
   removeToolState,
 } from '../../stateManagement/toolState.js';
+import debug from 'debug';
+
+const log = debug('cornerstoneTools')
+  .extend('eventDispatchers')
+  .extend('touchEventHandlers');
```
Comment (mikehazell) [inline src/eventDispatchers/touchEventHandlers/addNewMeasurement.js]: The pattern I'm trying to follow with log scoping is to roughly follow the file path.
There are some cases where the scope might be shared across files within a directory.

File: .eslintrc.js [hunk 2]
```diff
@@ -99,7 +99,7 @@ module.exports = {
     'no-caller': 'warn',
     'no-catch-shadow': 'warn',
     'no-confusing-arrow': 'warn',
-    'no-console': 'off',
+    'no-console': 'error',
```
Comment (dannyrb) [inline .eslintrc.js]: 🙌

File: src/thirdParty/registerItem.js [hunk 3]
```diff
@@ -15,19 +20,22 @@ import { lib } from '../lib.js';
...
+  }
 
-    if (overwrite) {
-      console.warn(`Overwriting ${type}/${name}`);
-    } else {
-      return;
-    }
+  if (alreadyRegistered) {
+    log('Overwriting %s', itemKey);
   }
```
Comment (mikehazell) [inline src/thirdParty/registerItem.js]: As I'm going through files, if I find a pattern that feels odd to me I'm doing minor refactors. This is my OCD coming out. If you'd rather I didn't do this please let me know. No hard feelings 😉

**Comment by mikehazell** [comment]
@dannyrb @JamesAPetts  - Checking this in early to see if you can see which direction I'm going  and give you an opportunity to steer if you want to.

This is a big library so I won't get to every file but I am trying to add consistent logging when I see a pattern.

File: .eslintrc.js (hunk 2 above)
Comment (mikehazell) [reply .eslintrc.js]: Haha, just makes them easy to find 😁

File: src/thirdParty/registerItem.js (hunk 3 above)
Comment (JamesAPetts) [reply src/thirdParty/registerItem.js]: It's all good, I wrote the whole plugin system in about 2 days during the 3.0 hype, any refactoring is very welcome!

File: src/thirdParty/registerSome.js [hunk 4]
```diff
@@ -13,9 +13,7 @@ import register from './register.js';
  * @returns {void}
  */
 export default function(items, overwrite = false) {
-  for (let i = 0; i < items.length; i++) {
-    const { type, name, item } = items[i];
-
+  items.forEach(({ type, name, item }) => {
```
Comment (JamesAPetts) [inline src/thirdParty/registerSome.js]: This is some pretty creative use of destructuring assignment, I like it.

**Comment by JamesAPetts** [comment]
I love this, very happy with the direction this is going!

File: package.json [hunk 5]
```diff
@@ -75,7 +75,8 @@
   },
   "dependencies": {
     "@babel/runtime": "7.1.2",
-    "cornerstone-math": "^0.1.7"
+    "cornerstone-math": "^0.1.7",
+    "debug": "^4.1.1"
```
Comment (dannyrb) [inline package.json]: Can you lock both of these to a specific version? (kill the `^`)

File: src/eventDispatchers/mouseEventHandlers/addNewMeasurement.js [hunk 6]
```diff
@@ -1,9 +1,15 @@
 import external from '../../externalModules.js';
 import { addToolState } from '../../stateManagement/toolState.js';
 import { moveHandle, moveNewHandle } from '../../manipulators/index.js';
+import debug from 'debug';
+
+const log = debug('cornerstoneTools')
```
Comment (dannyrb) [inline src/eventDispatchers/mouseEventHandlers/addNewMeasurement.js]: I'm not super familiar with this library, but it looks like we're create a debugger in multiple locations. Should we be creating a single `logger` instance in the store, and then extending it, as appropriate, in each module?

File: src/thirdParty/register.js [hunk 7]
```diff
@@ -48,20 +52,22 @@ export default function(type, name, item, overwrite = false) {
...
   if (!type) {
-    console.warn('The type must be given in order to register.');
+    log('The type must be given in order to register.');
 
     return false;
   }
 
   if (!name) {
-    console.warn(`The ${type} must have a name in order to register.`);
+    log(`The %s must have a name in order to register.`, type);
```
Comment (dannyrb) [inline src/thirdParty/register.js]: Is there a way to indicate that this is an error/warning state?

**Comment by dannyrb** [comment]
I like it (:
My questions so far are:

- How would we enable/disable logging entirely?
    - `init` setting?
    - WebPack to remove all `functions` when building for prod?
        - There is a plugin that does this for `console.log()`
- If we wanted to replace logging with a different library/implementation, but keep this same interface, how hard would it be? Do we need to implement something `adapter pattern`-y
- Is there a way for us to enable/disable logging for portions of the library? Or show different "log levels"?
    - `warnings`, `errors`, `debug/info`, `verbose`, `silly`

File: package.json (hunk 5 above)
Comment (mikehazell) [reply package.json]: Done

File: src/eventDispatchers/mouseEventHandlers/addNewMeasurement.js (hunk 6 above)
Comment (mikehazell) [reply src/eventDispatchers/mouseEventHandlers/addNewMeasurement.js]: I've done something like this now. I'm not worried about creating too many instances of the logger though. That's not likely to cause performance issues unless someone starts doing that inside a loop or event handler.

File: src/thirdParty/register.js (hunk 7 above)
Comment (mikehazell) [reply src/thirdParty/register.js]: Yep, done. I think there might be a conversation to be had / documentation to be written around how to use logging and when log, warn or error should be used.

**Comment by mikehazell** [comment]
> * How would we enable/disable logging entirely?
>   
>   * `init` setting?
>   * WebPack to remove all `functions` when building for prod?
>     * There is a plugin that does this for `console.log()`

Logging is disabled by default. The exceptions to this are: 
- `logger.error` will always print to console
- `logger.warn` will print to console in dev but be turned into a regular `logger.log` in prod.

> * If we wanted to replace logging with a different library/implementation, but keep this same interface, how hard would it be? Do we need to implement something `adapter pattern`-y

I have updated the implementation to enable this.

> * Is there a way for us to enable/disable logging for portions of the library? Or show different "log levels"?
>   
>   * `warnings`, `errors`, `debug/info`, `verbose`, `silly`

For "portions of the library" yes, for levels no. I think really this comes down to contributors having some restraint around what gets logged. We can of course add other log levels later if someone feels strongly about it.

`cornerstoneTools` now exports `enableLogger` and `disableLogger` methods.

`enableLogger` takes a scoping string with which you can specify scopes to be included / excluded.
Multiple scopes are comma separated. Excluding a scope is done by prefixing with `-`. You can use the `*` match against many scopes.

Some examples:

```js
// Enable logging
csTools.enableLogger();
// defaults to "cornerstoneTools:*" which will print all logs from the cornerstoneTools package

// only show logs from setToolMode.js
csTools.enableLogger("cornerstoneTools:store:setToolMode");

// exclude the eventDispatchers
csTools.enableLogger("cornerstoneTools:*,-cornerstoneTools:eventDispatchers:*");

// Disable logging
const prevSettings = csTools.disableLogger();
// This returns a string of the previous settings in case
// you wanted to toggle logging off and back on later.
csTools.enableLogging(prevSettings);
```

Under the hood, `debug` stores the config in `localStorage` so settings persist across a refresh.

**Comment by JamesAPetts** [comment]
Just had a look through, 3 green lights from me, this is really nice @mikehazell !

**Comment by mikehazell** [comment]
@dannyrb - I think I'm happy with this. Was there anything else you were after from me?

**Comment by dannyrb** [comment]
I love everything about this PR. Excellent work, @mikehazell! 🎉

**Pull request merged**